File: Oil Money project/Oil Money NOK.r
# Oil Money NOK - R Translation

# Load required libraries
library(tidyverse)
library(lubridate)
library(glmnet)
library(dplyr)
library(ggplot2)
library(tidyr)
library(urca)
library(tseries)
library(tidyverse)

# Load and prepare data
df <- read.csv("data/brent crude nokjpy.csv", check.names = FALSE)
colnames(df)[1] <- "date"
df$date <- as.Date(df$date, format = "%m/%d/%Y")

# Scatter plot NOK vs Brent
df %>%
  filter(date < as.Date("2018-04-24")) %>%
  ggplot(aes(x = brent, y = nok)) +
  geom_point(color = "#5f0f4e", size = 0.8) +
  labs(title = "NOK Brent Correlation", x = "Brent in JPY", y = "NOKJPY") +
  theme_minimal()

# Dual-axis base plot function (mimics Python twin axes)
dual_base_plot <- function(data, y1_col, y2_col, y1_label, y2_label, title,
                           col1 = "#34262b", col2 = "#cb2800") {
  # Set margins to allow right axis label
  par(mar = c(5, 4, 4, 4) + 0.1)
  # Plot first series
  plot(data$date, data[[y1_col]], type = "l", col = col1,
       xlab = "Date", ylab = y1_label, main = title)
  # Overlay second series with no axes
  par(new = TRUE)
  plot(data$date, data[[y2_col]], type = "l", col = col2,
       axes = FALSE, xlab = "", ylab = "")
  axis(side = 4, col = col2, col.axis = col2)
  mtext(y2_label, side = 4, line = 3, col = col2)
  legend("bottomright", legend = c(y1_label, y2_label),
         col = c(col1, col2), lty = 1, bty = "n")
}

# Use dual_base_plot for relationships
# NOK vs Interest Rate
dual_base_plot(
  df %>% filter(date < as.Date("2018-04-24")),
  y1_col = "nok", y2_col = "interest rate",
  y1_label = "NOKJPY", y2_label = "Interest Rate",
  title = "NOK vs Interest Rate"
)

# NOK vs Brent
dual_base_plot(
  df %>% filter(date < as.Date("2018-04-24")),
  y1_col = "nok", y2_col = "brent",
  y1_label = "NOKJPY", y2_label = "Brent in JPY",
  title = "NOK vs Brent",
  col2 = "#66ffff"
)

# NOK vs GDP (quarterly) 

# Get indices and values where quarterly GDP is available
gdp_idx    <- which(!is.na(df$`gdp yoy`))
gdp_dates  <- df$date[gdp_idx]
gdp_values <- df$`gdp yoy`[gdp_idx]
nok_values <- df$nok[gdp_idx]  # Get actual NOK values on the same GDP dates

# Prepare combined dataframe with matched frequency
gdp_plot_df <- data.frame(
  date = gdp_dates,
  nok = nok_values,
  gdp_yoy = gdp_values
) %>% filter(date < as.Date("2018-01-01"))

# Plot using dual_base_plot
dual_base_plot(
  gdp_plot_df,
  y1_col = "nok", y2_col = "gdp_yoy",
  y1_label = "NOKJPY", y2_label = "GDP YoY %",
  title = "NOK vs GDP"
)

# Linear regression (OLS)
x0 <- df %>% select(usd, gbp, eur, brent)
y  <- df$nok
train_idx <- df$date < as.Date("2017-04-25")
ols_model <- lm(y[train_idx] ~ ., data = x0[train_idx, ])
print(summary(ols_model))

#Elastic Net regression (glmnet)
# Define the cutoff date
cutoff_date <- as.Date("2017-04-25")

# Filter data before the cutoff date
x0_filtered <- as.matrix(df[df$date < cutoff_date, c("usd", "gbp", "eur", "brent")])
y_filtered <- df$nok[df$date < cutoff_date]

# Parameters to evaluate
l1_ratio <- 0.01  # alpha in glmnet (Python l1_ratio)
lambdas <- c(9.9999, 10, 10.0000001)  # lambda in glmnet, I set closest to the 10 because is the best value in Python (Python alpha) 

# Search for the best model
best_model <- NULL
best_mse <- Inf
best_alpha <- NA
best_lambda <- NA

set.seed(123)  # For reproducibility
fit <- cv.glmnet(
  x = x0_filtered,
  y = y_filtered,
  alpha = l1_ratio,
  lambda = lambdas,
  standardize = FALSE,  # <-- Avoid standardization (Python Elastic Net doesn't apply standardization)
  intercept = TRUE,
  nfolds = 5,
  maxit = 5000
)

if (min(fit$cvm) < best_mse) {
  best_mse <- min(fit$cvm)
  best_model <- fit
  best_alpha <- l1_ratio
  best_lambda <- fit$lambda.min
}

# Extract coefficients
coef_best <- coef(best_model, s = "lambda.min")
intercept <- coef_best[1]
coefs <- as.vector(coef_best[-1])

cat("Best alpha (l1_ratio):", best_alpha, "\n")
cat("Best lambda:", best_lambda, "\n")
cat("Intercept:", intercept, "\n")
cat("Coefficients:", round(coefs, 5), "\n")

# Compute fitted values and residuals for the full dataframe
df$sk_fit <- df$usd * coefs[1] +
  df$gbp * coefs[2] +
  df$eur * coefs[3] +
  df$brent * coefs[4] +
  intercept

df$sk_residual <- df$nok - df$sk_fit

# Signals generation logic
# Set thresholds based on 1 sigma of residuals before the cutoff date
residual_std <- sd(df$sk_residual[df$date < as.Date('2017-04-25')])
upper <- residual_std
lower <- -residual_std

# Create the signals dataframe starting from April 25, 2017
signals <- df %>%
  filter(date >= as.Date('2017-04-25')) %>%
  select(date, nok, usd, eur, gbp, brent, sk_fit, sk_residual)

# Rename 'sk_fit' column to 'fitted'
signals <- signals %>%
  rename(fitted = sk_fit)

# Add band levels and initialize signal columns
signals <- signals %>%
  mutate(
    stop_profit = fitted + 2 * upper,
    stop_loss = fitted + 2 * lower,
    upper = fitted + upper,
    lower = fitted + lower,
    signals = 0
  )

# Ensure signals dataframe has signal and cumulative signal columns
signals$signals <- 0
signals$cumsum <- 0  # to track cumulative signal exposure

# Loop row by row
for (j in 1:nrow(signals)) {
  
  # Entry rules
  if (signals$nok[j] > signals$upper[j]) {
    signals$signals[j] <- -1  # Short signal
  }
  
  if (signals$nok[j] < signals$lower[j]) {
    signals$signals[j] <- 1   # Long signal
  }
  
  # Recalculate cumulative signal count
  signals$cumsum <- cumsum(signals$signals)
  
  # Neutralize if cumulative exposure exceeds ±1
  if (signals$cumsum[j] > 1 || signals$cumsum[j] < -1) {
    signals$signals[j] <- 0
    #signals$cumsum <- cumsum(signals$signals)
  }
  
  # STOP PROFIT: if profit limit is reached
  if (signals$nok[j] > signals$stop_profit[j]) {
    signals$cumsum <- cumsum(signals$signals)
    signals$signals[j] <- -signals$cumsum[j] + 1
    signals$cumsum <- cumsum(signals$signals)
    break  # stop algorithm
  }
  
  # STOP LOSS: if loss limit is reached
  if (signals$nok[j] < signals$stop_loss[j]) {
    signals$cumsum <- cumsum(signals$signals)
    signals$signals[j] <- -signals$cumsum[j] - 1
    signals$cumsum <- cumsum(signals$signals)
    break  # stop algorithm
  }
}

# Ensure the 'date' column is of class Date
signals$date <- as.Date(signals$date)

# Create dataframes for LONG, SHORT, and Exit Point signals
long_signals <- signals %>%
  filter(signals > 0)

short_signals <- signals %>%
  filter(signals < 0)

exit_point <- signals %>%
  filter(date == as.Date("2017-12-20"))

# NOK signals positions plot
ggplot(signals, aes(x = date, y = nok)) +
  geom_line(color = "#594f4f", alpha = 0.5) +
  geom_point(data = long_signals, aes(x = date, y = nok),
             color = "#83af9b", shape = 24, fill = "#83af9b", size = 4) +
  geom_point(data = short_signals, aes(x = date, y = nok),
             color = "#fe4365", shape = 25, fill = "#fe4365", size = 4) +
  geom_point(data = exit_point, aes(x = date, y = nok),
             color = "#f9d423", shape = 8, size = 5, alpha = 0.8) +
  geom_vline(xintercept = as.Date("2017-11-15"), linetype = "dotted", color = "black") +
  labs(title = "NOKJPY Positions",
       x = "Date",
       y = "NOKJPY") +
  theme_minimal() +
  theme(panel.grid.minor = element_blank(),
        panel.grid.major = element_line(color = "gray90"),
        axis.title = element_text(face = "bold"),
        legend.position = "none")

# Fitted vs Actual plot
ggplot(signals, aes(x = date)) +
  geom_line(aes(y = fitted), color = "white", size = 1.2, alpha = 0.6) +
  geom_line(aes(y = nok), color = "black", size = 1.1, alpha = 0.8) +
  geom_ribbon(aes(ymin = lower, ymax = upper), fill = "#2a3457", alpha = 0.2) +
  geom_ribbon(aes(ymin = stop_loss, ymax = stop_profit), fill = "#720017", alpha = 0.1) +
  labs(title = "Fitted vs Actual", x = "Date", y = "NOKJPY") +
  theme_minimal()

# Normalized price trend plot 
# Normalize prices by dividing each by its first value and multiplying by 100
df_norm <- df %>%
  mutate(across(c(nok, usd, eur, gbp, brent), ~ . / .[1] * 100)) %>%
  select(date, nok, usd, eur, gbp, brent) %>%
  pivot_longer(cols = -date, names_to = "currency", values_to = "value")

# Custom colors
color_map <- c(
  nok   = "#ff8c94",  # Norwegian Krone
  usd   = "#9de0ad",  # US Dollar
  eur   = "#45ada8",  # Euro
  gbp   = "#f8b195",  # UK Sterling
  brent = "#6c5b7c"   # Brent Crude
)

# Custom labels for legend
label_map <- c(
  nok   = "Norwegian Krone",
  usd   = "US Dollar",
  eur   = "Euro",
  gbp   = "UK Sterling",
  brent = "Brent Crude"
)

# Create plot
ggplot(df_norm, aes(x = date, y = value, color = currency)) +
  geom_line(alpha = 0.9) +
  scale_color_manual(
    values = color_map,
    labels = label_map
  ) +
  labs(
    title = "Trend",
    x = "Date",
    y = "Normalized Price (Base = 100)",
    color = "Asset"
  ) +
  theme_minimal() +
  theme(
    panel.grid.minor = element_blank(),
    axis.title = element_text(face = "bold"),
    legend.position = "right"  # Puedes cambiar a "bottom", "left", etc.
  )

# Cointegration analysis
# Define the cutoff date
cutoff_date <- as.Date("2017-04-25")

# x2 is the predictor (eur), y is the dependent variable (nok)
x2 <- df$eur[df$date < cutoff_date]
y  <- df$nok[df$date < cutoff_date]

# OLS regression with intercept
ols_model <- lm(y ~ x2)
residuals <- resid(ols_model)

# ADF test on the residuals
adf_test <- ur.df(residuals, type = "none", selectlags = "AIC")
summary(adf_test)

# Show summary of the OLS regression
summary(ols_model)

# Performance analysis of signals strategy (Initial capital 2000, position size 100)
# Perform PnL (Profit and Loss) analysis
initial_capital <- 2000
position_size <- 100

# Create portfolio dataframe using the same index as signals
portfolio <- data.frame(date = signals$date)
rownames(portfolio) <- signals$date

# Calculate holding value: NOK price × cumulative position × position size
portfolio$holding <- signals$nok * signals$cumsum * position_size

# Calculate cash position: initial capital - cumulative cost of trades
portfolio$cash <- initial_capital - cumsum(signals$nok * signals$signals * position_size)

# Total asset value = holding + cash
portfolio$total_asset <- portfolio$holding + portfolio$cash

# Store signals for reference
portfolio$signals <- signals$signals

# Filter portfolio between 2017-10-01 and 2018-01-01
portfolio <- portfolio[portfolio$date > as.Date("2017-10-01") & portfolio$date < as.Date("2018-01-01"), ]

# Ensure 'date' column is of Date type
portfolio$date <- as.Date(portfolio$date)

# Compute standard deviation of total asset
asset_sd <- sd(portfolio$total_asset, na.rm = TRUE)

# Create subset for shaded region between 2017-11-20 and 2017-12-20
highlight_range <- portfolio %>%
  filter(date >= as.Date("2017-11-20") & date <= as.Date("2017-12-20")) %>%
  mutate(upper = total_asset + asset_sd,
         lower = total_asset - asset_sd)

# Create subsets for LONG and SHORT signal markers
longs <- portfolio %>% filter(signals > 0)
shorts <- portfolio %>% filter(signals < 0)

# Plot portfolio performance
ggplot(portfolio, aes(x = date, y = total_asset)) +
  geom_line(color = "#594f4f", alpha = 0.5, size = 0.7) +
  geom_point(data = longs, aes(x = date, y = total_asset),
             shape = 24, color = "#2a3457", fill = "#2a3457", size = 4, alpha = 0.5) +
  geom_point(data = shorts, aes(x = date, y = total_asset),
             shape = 25, color = "#720017", fill = "#720017", size = 6, alpha = 0.5) +
  geom_ribbon(data = highlight_range, aes(ymin = lower, ymax = upper),
              fill = "#547980", alpha = 0.2) +
  geom_vline(xintercept = as.Date("2017-11-15"), linetype = "dotted", color = "#ff847c") +
  annotate("text", x = as.Date("2017-12-20"),
           y = highlight_range$total_asset[highlight_range$date == as.Date("2017-12-20")] + asset_sd,
           label = "What if we use MACD here?", hjust = 0, size = 3.5) +
  labs(title = "Portfolio Performance",
       x = "Date", y = "Asset Value") +
  theme_minimal() +
  theme(panel.grid.minor = element_blank(),
        panel.grid.major = element_line(color = "gray90"),
        axis.title = element_text(face = "bold"),
        legend.position = "none")

# Oil_money_trading_backtest improved trading strategy optimised with gridsearch
# Import and run backtest module
source("oil_money_trading_backtest.r")
portfolio_data <- df %>% filter(date >= as.Date("2014-10-23"), date <= as.Date("2015-08-20"))
signals_bt   <- signal_generation(portfolio_data, "brent", "nok", oil_money)
portfolio_bt <- portfolio(signals_bt, "nok")
portfolio_bt$date <- portfolio_data$date
plot_signals(signals_bt, "nok")
graph_profit(portfolio_bt, "nok")

# Grid search for parameters
results <- expand.grid(h = 5:19, s = seq(0.3, 1.05, 0.05)) %>%
  mutate(return = map2_dbl(h, s, ~{
    sig <- signal_generation(df, "brent", "nok", oil_money, holding_threshold = .x, stop = .y)
    port <- portfolio(sig, "nok")
    tail(port$asset, 1) / head(port$asset, 1) - 1
  }))

# Trading returns distribution histogram (on NOK)
ggplot(results, aes(x = return * 100)) +
  geom_histogram(binwidth = 0.50, fill = "#f09e8c", color = "white") +
  labs(title = "Distribution of Return on NOK Trading", x = "Return (%)", y = "Frequency") +
  theme_minimal()

# Heatmap of returns changing Stop Profit/Loss and Holding Period
# Prepare data
return_df <- results %>%
  mutate(h = as.factor(h), s = as.factor(s)) %>%
  mutate(return = return * 100)  # convert to percentage

# Reverse factor levels of h (to invert y-axis)
return_df$h <- fct_rev(return_df$h)

# Plot
ggplot(return_df, aes(x = s, y = h, fill = return)) +
  geom_tile(color = NA) +  # no tile border
  scale_fill_gradientn(
    colors = c("#FFFFFF", "#ff8000", "#e63900", "#b30000", "#600000", "#000000"),
    name = "Return (%)",
    guide = guide_colorbar(
      direction = "vertical", 
      barwidth = 1, 
      barheight = 10,
      title.position = "top"
    )
  ) +
  coord_fixed() +  # makes tiles square
  labs(
    x = "Stop Profit/Loss",
    y = "Holding Period (days)",
    title = "Heatmap of Returns (%)"
  ) +
  theme_minimal(base_size = 15) +
  theme(
    panel.grid = element_blank(),           # remove grid
    legend.position = "right",              # legend on right
    legend.title.align = 0.5,               # center title
    legend.title = element_text(size = 10), # smaller title
    legend.text = element_text(size = 9),   # smaller labels
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

File: Oil Money project/Oil Money NOK.py

# coding: utf-8

# making graphs downloadable
import matplotlib
matplotlib.use('Agg')      # switch to a non-interactive backend
import matplotlib.pyplot as plt
plt.ioff()                 # turn off interactive mode

# In[1]:
# Importing libraries

import matplotlib.pyplot as plt
import statsmodels.api as sm
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.linear_model import ElasticNetCV as en 
from statsmodels.tsa.stattools import adfuller as adf



# In[2]:
# Load and preprocess the dataset

df=pd.read_csv('data/brent crude nokjpy.csv')
df.set_index(pd.to_datetime(df[list(df.columns)[0]]),inplace=True)
del df[list(df.columns)[0]]


# In[3]:
# Data visualitation and analysis
# Data from 2013-4-25 to 2017-4-25 as estimation horizon/training set
# Data from 2017-4-25 to 2018-4-25 as validation horizon/testing set

# Scatter plot of NOKJPY vs Brent before 2023 (correlation plot)
ax=plt.figure(figsize=(10,5)).add_subplot(111)
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.scatter(df['brent'][df.index<'2017-04-25'],df['nok'][df.index<'2017-04-25'],s=1,c='#5f0f4e')

plt.title('NOK Brent Correlation')
plt.xlabel('Brent in JPY')
plt.ylabel('NOKJPY')
plt.show()

# Dual-axis-plot function (we will use it to plot several graphs)
def dual_axis_plot(xaxis,data1,data2,fst_color='r',
                    sec_color='b',fig_size=(10,5),
                   x_label='',y_label1='',y_label2='',
                   legend1='',legend2='',grid=False,title=''):
    
    fig=plt.figure(figsize=fig_size)
    ax=fig.add_subplot(111)
    

    ax.set_xlabel(x_label)
    ax.set_ylabel(y_label1, color=fst_color)
    ax.plot(xaxis, data1, color=fst_color,label=legend1)
    ax.tick_params(axis='y',labelcolor=fst_color)
    ax.yaxis.labelpad=15

    plt.legend(loc=3)
    ax2 = ax.twinx()

    ax2.set_ylabel(y_label2, color=sec_color,rotation=270)
    ax2.plot(xaxis, data2, color=sec_color,label=legend2)
    ax2.tick_params(axis='y',labelcolor=sec_color)
    ax2.yaxis.labelpad=15

    fig.tight_layout()
    plt.legend(loc=4)
    plt.grid(grid)
    plt.title(title)
    plt.show()
    
# NOK vs Interest Rate plot
dual_axis_plot(df.index,df['nok'],df['interest rate'],
               fst_color='#34262b',sec_color='#cb2800',
               fig_size=(10,5),x_label='Date',
               y_label1='NOKJPY',y_label2='Norges Bank Interest Rate %',
               legend1='NOKJPY',legend2='Interest Rate',
               grid=False,title='NOK vs Interest Rate')

# NOK vs Brent in JPY plot
dual_axis_plot(df.index,df['nok'],df['brent'],
               fst_color='#4f2d20',sec_color='#3feee6',
               fig_size=(10,5),x_label='Date',
               y_label1='NOKJPY',y_label2='Brent in JPY',
               legend1='NOKJPY',legend2='Brent',
               grid=False,title='NOK vs Brent')
               
# NOK vs GDP yoy growth plot
# We need to convert NOK into quarterly data because the gdp is released quarterly
ind=df['gdp yoy'].dropna().index
dual_axis_plot(df.loc[ind].index,
               df['nok'].loc[ind],
               df['gdp yoy'].dropna(),
               fst_color='#116466',sec_color='#ff652f',
               fig_size=(10,5),x_label='Date',
               y_label1='NOKJPY',y_label2='Norway GDP YoY %',
               legend1='NOKJPY',legend2='GDP',
               grid=False,title='NOK vs GDP')


# Linear Regression (OLS)
x0=pd.concat([df['usd'],df['gbp'],df['eur'],df['brent']],axis=1)
x1=sm.add_constant(x0)
x=x1[x1.index<'2017-04-25']
y=df['nok'][df.index<'2017-04-25']

model=sm.OLS(y,x).fit()
print(model.summary(),'\n')


# In[4]:
# Elastic Net regression (ElasticNetCV function)
# Multicollinearity is observed in OLS so we use Elastic Net regression

m=en(alphas=[0.0001, 0.0005, 0.001, 0.01, 0.1, 1, 10],
     l1_ratio=[.01, .1, .5, .9, .99],  max_iter=5000).fit(x0[x0.index<'2017-04-25'], y)  
print(m.intercept_,m.coef_)

print("Best l1_ratio:", m.l1_ratio_) # Print Best l1_ratio and best alpha to reproduce same parameter in R
print("Best alpha:", m.alpha_)

# Elastic net estimation results:
# 3.79776228406 [ 0.00388958  0.01992038  0.02823187  0.00050092]


# In[5]:
# Calculate the fitted values of NOK
df['sk_fit']=(df['usd']*m.coef_[0]+df['gbp']*m.coef_[1]+
                 df['eur']*m.coef_[2]+df['brent']*m.coef_[3]+m.intercept_)


# In[6]:
# Calculate the residuals of NOK
df['sk_residual']=df['nok']-df['sk_fit']

# In[7]:
# Define trading signals based on residual standard deviation bands (based on Elastic Net results)
upper=np.std(df['sk_residual'][df.index<'2017-04-25'])
lower=-upper

signals=pd.concat([df[i] for i in ['nok', 'usd', 'eur', 'gbp', 'brent', 'sk_fit','sk_residual']], \
                  axis=1)[df.index>='2017-04-25']
signals['fitted']=signals['sk_fit']
del signals['sk_fit']

signals['upper']=signals['fitted']+upper
signals['lower']=signals['fitted']+lower
signals['stop profit']=signals['fitted']+2*upper
signals['stop loss']=signals['fitted']+2*lower
signals['signals']=0


# In[8]:


# Signal generation logic: 2 sigmas strategy with long/short/exit with Stop Loss and Stop Profit

index=list(signals.columns).index('signals')

for j in range(len(signals)):
    
    if signals['nok'].iloc[j]>signals['upper'].iloc[j]:
        signals.iloc[j,index]=-1  
          
    if signals['nok'].iloc[j]<signals['lower'].iloc[j]:
        signals.iloc[j,index]=1 
       
    signals['cumsum']=signals['signals'].cumsum()

    if signals['cumsum'].iloc[j]>1 or signals['cumsum'].iloc[j]<-1:
        signals.iloc[j,index]=0
  
    if signals['nok'].iloc[j]>signals['stop profit'].iloc[j]:         
        signals['cumsum']=signals['signals'].cumsum()
        signals.iloc[j,index]=-signals['cumsum'].iloc[j]+1
        signals['cumsum']=signals['signals'].cumsum()
        break

    if signals['nok'].iloc[j]<signals['stop loss'].iloc[j]:
        signals['cumsum']=signals['signals'].cumsum()
        signals.iloc[j,index]=-signals['cumsum'].iloc[j]-1
        signals['cumsum']=signals['signals'].cumsum()
        break


# In[9]:
# Plot trading signals on NOK chart
from datetime import datetime

ax=plt.figure(figsize=(10,5)).add_subplot(111)
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)

signals['nok'].plot(label='NOKJPY',c='#594f4f',alpha=0.5)
ax.plot(signals.loc[signals['signals']>0].index,
         signals['nok'][signals['signals']>0],
         lw=0,marker='^',c='#83af9b',label='LONG', markersize=10)
ax.plot(signals.loc[signals['signals']<0].index,
         signals['nok'][signals['signals']<0],
         lw=0,marker='v',c='#fe4365',label='SHORT', markersize=10)
ax.plot(pd.to_datetime('2017-12-20'),
         signals['nok'].loc['2017-12-20'],
         lw=0,marker='*',c='#f9d423', markersize=15, alpha=0.8,
         label='Potential Exit Point of Momentum Trading')

plt.axvline(datetime.strptime('2017/11/15', '%Y/%m/%d'), linestyle=':', c='k', label='Exit')
plt.legend()
plt.title('NOKJPY Positions')
plt.ylabel('NOKJPY')
plt.xlabel('Date')
plt.show()


# In[10]:
# Plot actual vs fitted with confidence bands
# After 2017/11/15, nokjpy price went skyrocketing

ax=plt.figure(figsize=(10,5)).add_subplot(111)
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)

signals['fitted'].plot(lw=2.5,label='Fitted',c='w',alpha=0.6)
signals['nok'].plot(lw=2,label='Actual',c='#04060f',alpha=0.8)
ax.fill_between(signals.index,signals['upper'],
                signals['lower'],alpha=0.2,label='1 Sigma',color='#2a3457')
ax.fill_between(signals.index,signals['stop profit'],
                signals['stop loss'],alpha=0.1,label='2 Sigma',color='#720017')

plt.legend(loc='best')
plt.title('Fitted vs Actual')
plt.ylabel('NOKJPY')
plt.xlabel('Date')
plt.show()


# In[11]:

# Brent crude drives NOK/JPY's short-term moves,
# while long-term shifts may stem from geopolitical events and U.S. policy changes.

# In[12]:
# Plot normalized price trends (for visual comparison)

ax=plt.figure(figsize=(10,5)).add_subplot(111)
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)

(df['nok']/df['nok'][0]*100).plot(c='#ff8c94',label='Norwegian Krone',alpha=0.9)
(df['usd']/df['usd'][0]*100).plot(c='#9de0ad',label='US Dollar',alpha=0.9)
(df['eur']/df['eur'][0]*100).plot(c='#45ada8',label='Euro',alpha=0.9)
(df['gbp']/df['gbp'][0]*100).plot(c='#f8b195',label='UK Sterling',alpha=0.9)
(df['brent']/df['brent'][0]*100).plot(c='#6c5b7c',label='Brent Crude',alpha=0.5)

plt.legend(loc='best')
plt.ylabel('Normalized Price by 100')
plt.xlabel('Date')
plt.title('Trend')
plt.show()


# In[13]:
# Stationarity test on residuals using ADF (Augmented Dickey-Fuller)

x2=df['eur'][df.index<'2017-04-25']
x3=sm.add_constant(x2)

model=sm.OLS(y,x3).fit()
ero=model.resid
adf_stat, pvalue, usedlag, nobs, crit_vals, icbest = adf(ero) # Test for stationarity

print(f"ADF Statistic : {adf_stat}")
print(f"p-value        : {pvalue}")
print(f"Used lag       : {usedlag}")
print(f"Number of obs. : {nobs}")
print("Critical values:")
for level, cv in crit_vals.items():
    print(f"  {level}: {cv}")
print(f"IC Best        : {icbest}")
print(model.summary())
print(adf(ero))
print(model.summary())

# We cant conclude any cointegration


# In[14]:
# Simulate trading strategy: backtest PnL (Initial capital 2000, each postiion of 10)
capital0=2000
positions=100
portfolio=pd.DataFrame(index=signals.index)
portfolio['holding']=signals['nok']*signals['cumsum']*positions
portfolio['cash']=capital0-(signals['nok']*signals['signals']*positions).cumsum()
portfolio['total asset']=portfolio['holding']+portfolio['cash']
portfolio['signals']=signals['signals']


# In[15]:
# Portfolio date range

portfolio=portfolio[portfolio.index>'2017-10-01']
portfolio=portfolio[portfolio.index<'2018-01-01']


# In[16]:
# Plot portfolio performance and signal points
ax=plt.figure(figsize=(10,5)).add_subplot(111)
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)

portfolio['total asset'].plot(c='#594f4f',alpha=0.5,label='Total Asset')
ax.plot(portfolio.loc[portfolio['signals']>0].index,portfolio['total asset'][portfolio['signals']>0],
         lw=0,marker='^',c='#2a3457',label='LONG',markersize=10,alpha=0.5)
ax.plot(portfolio.loc[portfolio['signals']<0].index,portfolio['total asset'][portfolio['signals']<0],
         lw=0,marker='v',c='#720017',label='The Big Short',markersize=15,alpha=0.5)
ax.fill_between(portfolio['2017-11-20':'2017-12-20'].index,
                 (portfolio['total asset']+np.std(portfolio['total asset']))['2017-11-20':'2017-12-20'],
                 (portfolio['total asset']-np.std(portfolio['total asset']))['2017-11-20':'2017-12-20'],
                 alpha=0.2, color='#547980')

plt.text(pd.to_datetime('2017-12-20'),
          (portfolio['total asset']+np.std(portfolio['total asset'])).loc['2017-12-20'],
          'What if we use MACD here?')
plt.axvline(datetime.strptime('2017/11/15', '%Y/%m/%d'),linestyle=':',label='Exit',c='#ff847c')
plt.legend()
plt.title('Portfolio Performance')
plt.ylabel('Asset Value')
plt.xlabel('Date')
plt.show()


# Though the model predicts well, profits arise mainly from momentum beyond stop thresholds, 
# suggesting it's more a trend-following strategy than statistical arbitrage.


# In[17]:
# Oil_money_trading_backtest improved trading strategy optimised
# Import trading backtest strategy
import oil_money_trading_backtest as om

dataset = df # We use this code for the short period: .loc['2014-10-23':'2015-08-20'], period [387:600] in the backtest
dataset.reset_index(inplace=True)

# Generate signals,monitor portfolio performance
# Plot positions and total asset
signals=om.signal_generation(dataset,'brent','nok',om.oil_money)
p=om.portfolio(signals,'nok')
om.plot(signals,'nok')
om.profit(p,'nok')

# Positive return but very low, just about 2%
# Strategy optimizer, holding period and Stop Loss/Profit
dic={}
for holdingt in range(5,20):
    for stopp in np.arange(0.3,1.1,0.05):
        signals=om.signal_generation(dataset,'brent','nok',om.oil_money,
                                     holding_threshold=holdingt,
                                    stop=stopp)
        
        p=om.portfolio(signals,'nok')
        dic[holdingt,stopp]=p['asset'].iloc[-1]/p['asset'].iloc[0]-1
     
profile=pd.DataFrame({'params':list(dic.keys()),'return':list(dic.values())})


# In[18]:

# Distribution of return plot
# Average return is 2%
# -6% and 6% as extreme values

ax=plt.figure(figsize=(10,5)).add_subplot(111)
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
profile['return'].apply(lambda x:x*100).hist(histtype='bar', \
                                            color='#f09e8c', \
                                            width=0.45,bins=20)
plt.title('Distribution of Return on NOK Trading')
plt.grid(False)
plt.ylabel('Frequency')
plt.xlabel('Return (%)')
plt.show()


# In[19]:

# Heatmap of holding period and Stop Loss/Profit return
matrix=pd.DataFrame(columns= \
                    [round(i,2) for i in np.arange(0.3,1.1,0.05)])

matrix['index']=np.arange(5,20)
matrix.set_index('index',inplace=True)

for i,j in profile['params']:
    matrix.at[i,round(j,2)]= \
    profile['return'][profile['params']==(i,j)].item()*100

for i in matrix.columns:
    matrix[i]=matrix[i].apply(float)


fig=plt.figure(figsize=(10,5))
ax=fig.add_subplot(111)
sns.heatmap(matrix,cmap='gist_heat_r',square=True, \
            xticklabels=3,yticklabels=3)
ax.collections[0].colorbar.set_label('Return(%) \n', \
                                     rotation=270)
plt.xlabel('\nStop Loss/Profit (points)')
plt.ylabel('Position Holding Period (days)\n')
plt.title('Profit Heatmap\n',fontsize=10)
plt.style.use('default')

# Stop profit/loss point doesn't really affect (Best: from 0.6 to 1.05)
# More correlated with the length of holding period (Best: 9 trading days)

#In[20]
# Download all outputted figures in the file
import os

base_dir   = 'NOK Data/'
out_folder = os.path.join(base_dir, 'py_graphs')
os.makedirs(out_folder, exist_ok=True)

for idx, num in enumerate(plt.get_fignums(), start=1):
    fig = plt.figure(num)
    fig.savefig(
        os.path.join(out_folder, f'figure_{idx}.png'),
        dpi=300, bbox_inches='tight'
    )
plt.close('all')
# %%